tools/distpartitioning/dist_lookup.py

```diff
@@ -0,0 +1,289 @@
+
+        global_nids <-> partition-ids mappings are deterministically  distributed across 
+        all the participating processes, within the service. A contiguous global-nids
+        (ntype-ids, per-type-nids) are stored within each process and this is determined 
+        by the total no. of nodes of a given ntype-id and the rank of the process.
+
+        Process, where the global_nid <-> partition-id mapping is stored can be easily computed
+        as described above. Once this is determined we perform an alltoallv to send the request.
+        On the receiving side, each process receives a set of global_nids and retrieves corresponding
+        partition-ids using locally stored lookup tables. It builds responses to all the other
+        processes and performs alltoallv.
+
+        Once the response, partition-ids, is received, they are re-ordered corresponding to the 
+        incoming global-nids order and returns to the caller.
+
+        Parameters:
+        -----------
+        self : instance of this class
+            instance of this class, which is passed by the runtime implicitly
+        global_nids : numpy array
+            an array of global node-ids for which partition-ids are to be retrieved by 
+            the distributed lookup service.
+
+        Returns:
+        --------
+        list of integers : 
+            list of integers, which are the partition-ids of the global-node-ids (which is the
+            function argument)
+        '''
+
+        #find the process where global_nid --> partition-id(owner) is stored. 
+        ntype_ids, type_nids = self.id_map(global_nids)
+        ntype_ids, type_nids = ntype_ids.numpy(), type_nids.numpy()
+        assert len(ntype_ids) == len(global_nids)
+
+        #For each node-type, the per-type-node-id <-> partition-id mappings are
+        #stored as contiguous chunks by this lookup service. 
+        #The no. of these mappings stored by each process, in the lookup service, are
+        #equally split among all the processes in the lookup service, deterministically
+        service_owners = self.ntype_count[ntype_ids]
+        service_owners = np.ceil(service_owners/self.world_size).astype(np.int64)
+        service_owners = np.floor_divide(type_nids, service_owners).astype(np.int64)
+
+        #now `service_owners` is a list of ranks (process-ids) which own the corresponding
+        #global-nid <-> partition-id mapping.
+
+        #Split the input global_nids into a list of lists where each list will be 
+        #sent to the respective rank/process
+        #We also need to store the indices, in the indices_list, so that we can re-order
+        #the final result (partition-ids) in the same order as the global-nids (function argument)
+        send_list = []
+        indices_list = []
+        for idx in range(self.world_size):
+            cond = service_owners == idx
+            idxes = np.where(cond)
+            ll = global_nids[idxes[0]]
```

Comment:
why don't you write `global_nids[service_owners == idx]`?

Reply:
Agreed. Changed.